sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -2007,7 +2007,14 @@ case class Concat(children: Seq[Expression]) extends Expression {
     }
   }
 
-  override def dataType: DataType = children.map(_.dataType).headOption.getOrElse(StringType)
+  override def dataType: DataType = {
+    val dataTypes = children.map(_.dataType)
+    dataTypes.headOption.map {
+      case ArrayType(et, _) =>
+        ArrayType(et, dataTypes.exists(_.asInstanceOf[ArrayType].containsNull))
+      case dt => dt
+    }.getOrElse(StringType)
+  }
```

Comment:
Can't we handle this case in type coercion (analysis phase)?

Reply:
Actually, `Concat` for array type has the type coercion to add casts to make all children the same type, but we also have the optimization `SimplifyCasts` to remove unnecessary casts which might remove casts from arrays not contains null to arrays contains null ([optimizer/expressions.scala#L611](https://github.com/apache/spark/blob/d87a8c6c0d1a4db5c9444781160a65562f8ea738/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala#L611)).

E.g., `concat(array(1,2,3), array(4,null,6))` might generate a wrong data type during the execution.

Reply:
Added a test to show the wrong nullability.

Reply:
Aha, I see. But,  I just have a hunch that `SimplifyCasts` cannot simplify array casts in some cases?, e.g., this concat case.  Since we basically cannot change semantics in optimization phase, I feel a little weird about this simplification.

Anyway, I'm ok with your approach because I can't find a better & simpler way to solve this in analysis phase... Thanks!

Reply:
Hmm, that also makes sense. I'm not sure we can remove the simplification, though. cc @gatorsmile @cloud-fan

Reply:
I think `SimplifyCasts` is valid, but the concat type coercion rule is sub-optimal. We should not force all the children to have same `containsNull`, which may add unnecessary null checks inside `Concat`.

BTW are there more expression like `Concat` that we need to fix the `containsNull`?

Reply:
Currently we've found `CaseWhen` and `If` as at #21687. Seems like there are no other expressions in the collection functions at a glance.

Reply:
So far, we've identified also ```CaseWhen``` and  ```If``` discussed [here](https://github.com/apache/spark/pull/21687). I've just noticed that ```Coalesce``` looks also suspicious.

What is the key purpose of ```SimplifyCasts```? to remove an extra expression node or avoid casts from between to indentical types? If the second option is the purpose, what about changing ```SimplifyCasts``` rule to start replacing  ```Cast``` with a new dummy cast expression that will hold only a target data type and won't perform any casting?

Reply:
Yeah, `Coalesce` should be fixed, and `Least` and `Greatest` are also suspicious.

Reply:
Oh, see. In that case, it would be nice to introduce a method that will resolve the output DataType and merges ```nullable```/```containNull``` flags of non-primitive types recursively for such expressions. For the most cases we could encapsulate the function into a new trait. WDYT?

Reply:
SGTM

Reply:
+1 for the @mn-mikke idea